### ISSUE TITLE
screen reader logical reading order issue in dynamic page layout header section

### ISSUE THREAD

**Issue by anuvenkatesh1** [SAP/ui5-webcomponents-react#3287]
### Discussed in https://github.com/SAP/ui5-webcomponents-react/discussions/3267

<div type='discussions-op-text'>

<sup>Originally posted by **anuvenkatesh1** July 19, 2022</sup>
One of our pages use the [dynamic page layout component](https://sap.github.io/ui5-webcomponents-react/?path=/docs/layouts-floorplans-dynamicpage--default-story), which has the object header and content part just above the table. 
The reading order is Title & Status (1, 2) -> all buttons -> then the other details (location/availability etc) are read in screen reader Browse mode. May be in the context of this test page is ok. 
 

However, in the context of our page ( its about goals that we create and this page has the goal details. The object page header summarizes the details) and so it would make sense for a screen reader user to read all the details on the left first before moving on to the actions that can be taken on them ie in the context of the above dynamic page layout component, we want the title/status/updated/location/availability etc to be read first, before reading the buttons.
Please note that in our page we do not have the "collapse" button to collapse the header content area.

Below are the screenshots. 

Due to current structure of the object header, unable to change the reading order. What's the workaround? How to achieve this?
![image](https://user-images.githubusercontent.com/32471459/180473002-c70f55e2-0997-4d69-8521-37407114a624.png)

![image](https://user-images.githubusercontent.com/32471459/180472883-4fbf0d39-c216-404b-89b1-9f128f415e8e.png)

</div>

**Comment by don-obrien** [comment]
@MarcusNotheis @Lukas742 
Can we get an update on this one?  I see the bug tag.   Is this a bug, or does the reporter need advice on implementation?

**Comment by Lukas742** [comment]
Hi @don-obrien 

I didn't find the time yet to take a deeper look into this, but I plan to check it out tomorrow. If we need more input from the reporter we will ask for the necessary information here and add the "author action" label.

**Comment by Lukas742** [comment]
Hi @anuvenkatesh1 

I could not reproduce the reading order in your first example. Per default, so without setting e.g. `tab-index` or programmatically changing the focused element, only the active items are being focused and therefore read out. 

What screen reader are you using and is this is there an option I have to enable somewhere to get the same behavior as you?

**Comment by GongRichard** [comment]
Hi @Lukas742,
You can use this page https://github.wdf.sap.corp/pages/xweb/team-goal/app/?templateId=1&mock#/detail?id=1 to reproduce this issue. I think @anuvenkatesh1 used JAWS as screen reader.
Thanks,
Richard

**Comment by Lukas742** [comment]
Hi @GongRichard 

I could reproduce the described behavior using JAWS, but unfortunately I don't think there is much we can do. It seems like JAWS is reading the entries according to the DOM structure, which is the expected behavior in my opinion. Furthermore, it's the same behavior as with SAPUI5.  As I'm not that familiar with JAWS I'm not sure if you can configure that behavior by e.g. setting attributes or changing the `tab-index`. However, this would then also have to be changed by the app developer, as all elements can be accessed via props or by passing the attributes/props to the component (e.g. `Title`) itself.

**Comment by don-obrien** [comment]
Is this still in author action?  Do you need more info from @GongRichard ?

**Comment by Lukas742** [comment]
Hi @don-obrien 

please refer to my last comment.

**Comment by GongRichard** [comment]
Hi @Lukas742 ,
I don't think this is a good solution for us. From fiori guideline https://experience.sap.com/fiori-design-web/dynamic-page-layout/, the title prop only puts title information. If we put all header content in the title, I think there will cause some other issues, for example, action buttons alignment with title content.

Hi @anuvenkatesh1,
From fiori design guideline, only title and sub-title were put in header title area. Other infos were put in header content area. Our design meets the fiori guideline. I think we can hold on this issue and see whether there has customer will report a11y issue for this. 
<img width="1124" alt="image" src="https://user-images.githubusercontent.com/13098801/189045216-5094f597-37b9-4836-8855-ceb8686fc9d3.png">
<img width="1052" alt="image" src="https://user-images.githubusercontent.com/13098801/189045271-6c08aeb6-c906-4da2-967b-ed412038bdf4.png">

Thanks,
Richard

**Comment by Lukas742** [comment]
Hi @GongRichard 

what exactly do you think is not a good solution? Our DynamicPage is implemented according to the UX guidelines (except for the Summary Line feature, for which you are welcome to create a feature request) or did you find something that is not?

Also, what do you mean with "put all header content in the title"? - How do you define "header content" and "title"?
If you mean the `header` prop of the `DynamicPageTitle` component with "title", then this is something I never mentioned. The `DynamicPageTitle` receives different props where you either can pass in components (like `Title` in the `header` prop), which logically can be fully controlled by you, or you can control the internally used components like the two `Toolbar`s via props like `actionsToolbarProps`.

**Comment by GongRichard** [comment]
Hi @Lukas742 ,
Now we use DynamicPageTitle component, do you mean put all header content including title and subtitle to header prop of DynamicPageTitle?
<img width="453" alt="image" src="https://user-images.githubusercontent.com/13098801/189062619-3da7c694-fb7a-4537-8e1a-7e5b1dfeaabf.png">

Thanks,
Richard

**Comment by Lukas742** [comment]
No, that's not what I want you to do.
You need to use the DynamicPage in the intended way and set whatever customization you need on top of the props you pass into. Also, you didn't elaborate on what you mean with "title" or "header content" so until that is cleared up I unfortunately can't help you, but I'll try to summarize my findings and suggestions once more:

- The reading order of the JAWS screen reader is defined by the structure of the DOM, so we can't change that in our implementation.
- I don't know if JAWS offers an option to control this behavior e.g. via an attribute you can set to the element, or if it changes according to `tab-index`. If this is possible, then you can add these attributes to each of the components you pass into the `DynamicPageTitle` or `DynamicPageHeader`.
Small example with custom `tab-index` for the heading and each button:
```jsx
<DynamicPageTitle
  actions={
    <>
      <Button tabIndex={/*some value*/}>
        Edit
      </Button>
      <Button tabIndex={/*some value*/}>
        Delete
      </Button>
    </>
  }
  header={<Title tabIndex={/*some value*/}>Header Title</Title>}
/>
```
- As stated in our [documentation](https://sap.github.io/ui5-webcomponents-react/?path=/docs/layouts-floorplans-dynamicpage--default-story), you should only use the `DynamicPageTitle` component as value for the `headerTitle` prop of the `DynamicPage` and only `DynamicPageHeader` for the `headerContent` prop to preserve the intended design.

**Comment by GongRichard** [comment]
Hi @Lukas742 ,
Thanks for your suggestions. Change tab-index will cause axe error. I think the only way is to change the DOM structure. 
As I replied in the above thread, I also think our implementation meets fiori guideline. I suggested to hold on this issue until customer reports this as a11y issue to us.
Thanks,
Richard

**Comment by don-obrien** [comment]
@Lukas742 , an SFSF customer has complained about the same thing.  Can you please revisit this and plan for addressing it?

**Comment by Lukas742** [comment]
Hi @don-obrien 

as we are waiting for the outcome of the internal discussion, we will keep this topic closed until the decisions are made.
